pkg/cloudprovider/providers/azure/controllerCommon.go

```diff
@@ -0,0 +1,403 @@
+	}
+
+	defer resp.Body.Close()
+	if err != nil || resp.StatusCode != 200 {
+		return nil, getRestError(fmt.Sprintf("Get ARM VM: %s", armVMName), err, 200, resp.StatusCode, resp.Body)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return body, nil
+}
+
+func (c *controllerCommon) getToken() (token string, err error) {
+	if c.aadToken != "" && time.Now().UTC().Sub(c.expiresOn).Seconds() <= 10 {
+		// token cached and is valid.
+		return c.aadToken, nil
+	}
+
+	apiURL := c.tokenEndPoint
+	resource := fmt.Sprintf(aadTokenEndPointPath, c.tenantID)
+	// create form urlencoded post data
+	formData := url.Values{}
+
+	formData.Add("grant_type", "client_credentials")
+	formData.Add("client_id", c.clientID)
+	formData.Add("client_secret", c.clientSecret)
+	formData.Add("resource", c.aadResourceEndPoint)
+
+	urlStr := ""
+	u := &url.URL{}
+	if u, err = url.ParseRequestURI(apiURL); err != nil {
+		return "", err
+	}
+
+	u.Path = resource
+	urlStr = fmt.Sprintf("%v", u)
+	client := &http.Client{}
+
+	r, err := http.NewRequest("POST", urlStr, bytes.NewBufferString(formData.Encode()))
+	if err != nil {
+		return "", err
+	}
+
+	// add headers
+	r.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	r.Header.Add("Content-Length", strconv.Itoa(len(formData.Encode())))
+	resp, err := client.Do(r)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	payload, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	c.aadToken, c.expiresOn, err = parseAADToken(payload)
+	return c.aadToken, err
+}
+
+func parseAADToken(payload []byte) (string, time.Time, error) {
+	var f interface{}
+	var sToken string
+	var expiresOn time.Time
+	var expiresSec int
+	var ok bool
+	var expires string
+
+	if err := json.Unmarshal(payload, &f); err != nil {
+		return "", expiresOn, err
+	}
+
+	fragment, ok := f.(map[string]interface{})
+	if !ok {
+		return "", expiresOn, fmt.Errorf("convert vmData to map error")
+	}
+	if sToken, ok = fragment["access_token"].(string); ok != true {
+		return "", expiresOn, fmt.Errorf("Disk controller (ARM Client) cannot parse AAD token - access_token field")
+	}
+	if expires, ok = fragment["expires_on"].(string); ok != true {
+		return "", expiresOn, fmt.Errorf("Disk controller (ARM Client) cannot parse AAD token - expires_on field")
+	}
+
+	expiresSec, _ = strconv.Atoi(expires)
+	// expires_on is seconds since 1970-01-01T0:0:0Z UTC
+	expiresOn = time1970.UTC().Add(time.Duration(expiresSec) * time.Second)
+
+	return sToken, expiresOn, nil
+}
+
+// Creats a slice  luns based on the VM size, used the static map declared on this package
+func getLunMapForVM(vmSize string) []bool {
+	count, ok := dataDisksPerVM[vmSize]
+	if !ok {
+		glog.Warningf("azureDisk - VM Size %s found no static lun count will use default which  %v", vmSize, defaultDataDiskCount)
+		count = defaultDataDiskCount
+	}
+
+	m := make([]bool, count)
+	return m
+}
+
+var mutex = &sync.Mutex{}
```

Comment:
this won't work if the lun you get this time is not immediately consumed and filled by azure compute. You need a lock to protect findEmptyLun and attach disk.